### ISSUE TITLE
-torcontrol does not resolve a hostname if it is given

### ISSUE THREAD

**Issue by jrawsthorne** [bitcoin/bitcoin#22236]
**Expected behavior**

-torcontrol should try to resolve a hostname if it is given

**Actual behavior**

The hostname is passed as an ip so the connection fails

```
2021-06-13T20:35:54Z tor: Error parsing socket address mynode.local:9051
2021-06-13T20:35:54Z tor: Initiating connection to Tor control port mynode.local:9051 failed
```

**To reproduce**

Add ```127.0.0.1 mynode``` to /etc/hosts and pass -torcontrol=mynode:9051 as an argument. The connection fails. Try with -torcontrol=127.0.0.1:9051 and it works.

**System information**

Bitcoin Core 0.21.1 from https://bitcoincore.org/bin/bitcoin-core-0.21.1/bitcoin-0.21.1-x86_64-linux-gnu.tar.gz
Ubuntu 18.04

**Comment by prayank23** [comment]
What if you replace `127.0.0.1 mynode` with `127.0.0.1 mynode.local` in /etc/hosts and pass -torcontrol=mynode.local:9051 as an argument?

**Comment by jrawsthorne** [comment]
Still get the same error
```
2021-06-13T20:35:54Z tor: Error parsing socket address mynode.local:9051
2021-06-13T20:35:54Z tor: Initiating connection to Tor control port mynode.local:9051 failed
```

**Comment by jrawsthorne** [comment]
It works for -proxy. Looks like it calls Lookup() first to resolve the hostname
https://github.com/bitcoin/bitcoin/blob/d75a1df6178aabf6f19271786e48bc5d57eddb54/src/init.cpp#L1218-L1226

For -torcontrol it passes the hostname straight through
https://github.com/bitcoin/bitcoin/blob/d75a1df6178aabf6f19271786e48bc5d57eddb54/src/torcontrol.cpp#L579-L584

**Comment by prayank23** [comment]
> It works for -proxy. Looks like it calls Lookup() first to resolve the hostname. For -torcontrol it passes the hostname straight through

Yes we can try using lookup for torcontrol as well. However I noticed a weird thing: hostname works in Ubuntu but fails in Fedora.

```
proxy=localhost:9050
```

![image](https://user-images.githubusercontent.com/13405205/121824358-5581e100-ccc9-11eb-8347-f1ea500e77b9.png)

**Comment by prayank23** [comment]
Maybe I need to remove `::1` from hosts file.

**Comment by juxeii** [comment]
I still have this problem in 22.0.
Tor runs in a docker container and also bitcoind in a separete container(both bridged).
`proxy="host.docker.internal:9050"` resolves fine to the docker host `192.168.1.11:9050`
but 
`torcontrol=host.docker.internal:9051` fails with 
```
2022-04-07T06:26:53Z torcontrol thread start
2022-04-07T06:26:53Z tor: Error parsing socket address host.docker.internal:9051
2022-04-07T06:26:53Z tor: Initiating connection to Tor control port host.docker.internal:9051 failed
2022-04-07T06:26:53Z torcontrol thread exit
```
What can I do here when both Tor and bitcoind run in containers?

**Comment by caheredia** [comment]
> I still have this problem in 22.0. Tor runs in a docker container and also bitcoind in a separete container(both bridged). `proxy="host.docker.internal:9050"` resolves fine to the docker host `192.168.1.11:9050` but `torcontrol=host.docker.internal:9051` fails with
> 
> ```
> 2022-04-07T06:26:53Z torcontrol thread start
> 2022-04-07T06:26:53Z tor: Error parsing socket address host.docker.internal:9051
> 2022-04-07T06:26:53Z tor: Initiating connection to Tor control port host.docker.internal:9051 failed
> 2022-04-07T06:26:53Z torcontrol thread exit
> ```
> 
> What can I do here when both Tor and bitcoind run in containers?

@juxeii were you able to resolve this? I'm running the same setup as you.

**Comment by caheredia** [comment]
> 

@juxeii nevermind. I fixed by assigning fixed IPs to tor and bitcoind, using this example https://forums.docker.com/t/static-ip-on-docker-containers/110412/3
